freeipa/src/main/java/com/sequenceiq/freeipa/service/stack/FreeIpaRecommendationService.java

```diff
@@ -0,0 +1,90 @@
+import com.sequenceiq.freeipa.api.v1.freeipa.stack.model.create.VmTypeResponse;
+import com.sequenceiq.freeipa.converter.cloud.CredentialToExtendedCloudCredentialConverter;
+import com.sequenceiq.freeipa.converter.instance.VmTypeToVmTypeResponseConverter;
+import com.sequenceiq.freeipa.dto.Credential;
+import com.sequenceiq.freeipa.service.CredentialService;
+import com.sequenceiq.freeipa.service.stack.instance.DefaultInstanceTypeProvider;
+
+@Service
+public class FreeIpaRecommendationService {
+
+    @Inject
+    private CredentialService credentialService;
+
+    @Inject
+    private CloudParameterService cloudParameterService;
+
+    @Inject
+    private CredentialToExtendedCloudCredentialConverter extendedCloudCredentialConverter;
+
+    @Inject
+    private DefaultInstanceTypeProvider defaultInstanceTypeProvider;
+
+    @Inject
+    private VmTypeToVmTypeResponseConverter vmTypeConverter;
+
+    public FreeIpaRecommendationResponse getRecommendation(String credentialCrn, String region, String availabilityZone) {
+        Credential credential = credentialService.getCredentialByCredCrn(credentialCrn);
+        String defaultInstanceType = defaultInstanceTypeProvider.getForPlatform(credential.getCloudPlatform());
+        Set<VmTypeResponse> availableVmTypes = getAvailableVmTypes(region, availabilityZone, credential, defaultInstanceType);
+        return new FreeIpaRecommendationResponse(availableVmTypes, defaultInstanceType);
+    }
+
+    private Set<VmTypeResponse> getAvailableVmTypes(String region, String availabilityZone, Credential credential, String defaultInstanceType) {
+        CloudVmTypes vmTypes = cloudParameterService.getVmTypesV2(extendedCloudCredentialConverter.convert(credential),
+                region, credential.getCloudPlatform(), CdpResourceType.DEFAULT, Maps.newHashMap());
+
+        Set<VmType> availableVmTypes = Collections.emptySet();
+        if (StringUtils.isNotBlank(availabilityZone)) {
+            availableVmTypes = vmTypes.getCloudVmResponses().get(availabilityZone);
+        } else if (vmTypes.getCloudVmResponses() != null && !vmTypes.getCloudVmResponses().isEmpty()) {
+            availableVmTypes = vmTypes.getCloudVmResponses().values().iterator().next();
+        }
+
+        Optional<VmType> defaultVmType = availableVmTypes.stream()
+                .filter(vmType -> defaultInstanceType.equals(vmType.value()))
+                .findAny();
+        Set<VmType> filteredVmTypes = filterVmTypes(availableVmTypes, defaultVmType);
+        return filteredVmTypes.stream()
+                .map(vmType -> vmTypeConverter.convert(vmType))
+                .collect(Collectors.toSet());
+    }
+
+    private Set<VmType> filterVmTypes(Set<VmType> availableVmTypes, Optional<VmType> defaultVmType) {
+        return availableVmTypes.stream()
+                .filter(vmType -> filterVmTypeSmallerThanDefault(vmType, defaultVmType))
+                .collect(Collectors.toSet());
+    }
```

Comment:
this can be smiplified:
```
        Optional<VmType> defaultVmType = availableVmTypes.stream()
                .filter(vmType -> defaultInstanceType.equals(vmType.value()))
                .findAny();
        return availableVmTypes.stream()
                .filter(vmType -> filterVmTypeSmallerThanDefault(vmType, defaultVmType))
                .map(vmType -> vmTypeConverter.convert(vmType))
                .collect(Collectors.toSet());
```